### ISSUE TITLE
LiveRide: Directions sometimes confusing

### ISSUE THREAD

**Issue by oliverlockwood** [cyclestreets/android#276]
I don't think this is a LiveRide bug, more of a routing issue.

For example, there have been many times that I've approached a crossroads, perhaps going to cross a more major road, and the instruction is given to "turn left".  I duly do so, only to be told to "continue zero metres, get ready to turn right".

@mvl22 is this something you're aware of server side?

If not, or if the fix isn't likely to happen imminently, I might put in a workaround on the app side, e.g. to look ahead and merge instructions if the interval is below 10 metres or so... because this has been annoying me for a while.

**Comment by jezhiggins** [comment]
Oh man, I've thought about this a lot - my name for this is segment melding.  There are some cases when the road is in multiple segments but in reality it's just one continuous road, and those are kind of ok. The other cases, which really can be a drag, are the turn left, go 20 metres, turn right occurrences.  I tried to come up with a scheme that handled both but, with the benefit of some distance, I probably was overthinking it. Have a go.

**Comment by mvl22** [comment]
Can you send some specific example routes if possible. Ideally we want to combine traversal through a junction into the overall direction movement, not individual pieces.

**Comment by oliverlockwood** [comment]
Hi @mvl22, here's a number of specific examples, narrowed down from much longer routes I have in my history.
-  63090075 (10 metres on Bradenham Lane)
-  63090089 (9 metres on Pebble Hill)
-  63090095 (10 metres on High Street, A338)

In each case, you can go to https://www.cyclestreets.net/api/journey.json?plan=balanced&itinerary=<itineraryId>&key=<redacted> to see the raw JSON with the relevant segments; if you load those routes in the app, you'll be able to navigate through the segments and see where there's a problem.

I can find more if necessary but that should give you the general idea.

Watch this space for a PR which demonstrates what the voice notifications sound like in this circumstance.

**Comment by si-the-pie** [comment]
Yes, I agree that the root cause of this is problem lies with the output from the routing engine, which could do more to help.

Look at this route near Blackfriars Bridge in Central London. It starts in Queen Victoria Street and basically **turns left** onto Blackfriar's Bridge, but it is drawn in OSM as a complex wiggle and this is regurgitated in the complex and detailed routing instructions:

https://www.cyclestreets.net/journey/63112198/#fastest

The current server output does some _segment melding_ before producing the route listing, such as combining ways that are of the same highway type and having exactly the same name. But wherever there is a junction - even the artificial one at https://www.openstreetmap.org/node/5177397835 the listing will give a turn instruction.

There is also quite a bit of processing that handles roundabouts to try and work out which exit number to use.
Other cases like 'steps, bridge, steps' should be converted to 'go over bridge'.

Producing human friendly route listings then is the challenge, and that should probably be done on the server side.

https://github.com/cyclestreets/cyclestreets/issues/512

**Comment by oliverlockwood** [comment]
It's good to know that this is something you're looking to fix server-side.

In the short term (since my experience is that server-side fixes can take some time) I'm going to ameliorate some of these experiences within the app - PR to come shortly.
